File: flatten_bin_tree.cpp
//TC: O(N)
//SC: O(N)

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void flatten(TreeNode* root) {
        
        if(root == NULL){
            return;
        }
        
        stack<TreeNode*> s;
        
        s.push(root);
        
        while(!s.empty()){
            TreeNode* curr = s.top();
            s.pop();
            
            if(curr->right!=NULL){
                s.push(curr->right);
            }
            
            if(curr->left!=NULL){
                s.push(curr->left);
            }
            
            curr->left = NULL;
            
            if(!s.empty()){
                curr->right = s.top();
            } else {
                curr->right = NULL;
            }
            
        }
        
        return;
        
        
    }
};

File: flattenbintree.py
"""114. Flatten Binary Tree to Linked List
Time Complexity: O(n)
Space Complexity: O(n)"""

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def flatten(self, root: TreeNode) -> None:
        """
        Do not return anything, modify root in-place instead.
        """
        if root is None:
            return None
        stack = [root]
        curr = root
        
        while stack: #root needed or not?
            curr = stack.pop()
            
            if curr.right:
                stack.append(curr.right)
            if curr.left:
                stack.append(curr.left)
                
            if curr and stack:
                curr.left = None
                curr.right = stack[-1]
            
        return root
            
            
            